pkg/querysharding/analyzer.go

```diff
@@ -0,0 +1,106 @@
+package querysharding
+
+import (
+	"fmt"
+
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+// QueryAnalyzer is an analyzer which determines
+// whether a PromQL Query is shardable and using which labels.
+type QueryAnalyzer struct{}
+
+var nonShardableFuncs = []string{
+	"label_join",
+	"label_replace",
+}
+
+// NewQueryAnalyzer creates a new QueryAnalyzer
+func NewQueryAnalyzer() *QueryAnalyzer {
+	return &QueryAnalyzer{}
+}
+
+// Analyze analyzes a query and returns a QueryAnalysis.
+
+// Analyze uses the following algorithm:
+// * if a query has subqueries, such as label_join or label_replace,
+//   then treat the query as non shardable.
+// * if the query's root expression has grouping labels,
+//   then treat the query as shardable by those labels.
+// * if the query's root expression has no grouping labels,
+//   then walk the query and find the least common labelset
+//   used in grouping expressions. If non-empty, treat the query
+//   as shardable by those labels.
+// * otherwise, treat the query as non-shardable.
+func (a *QueryAnalyzer) Analyze(query string) (QueryAnalysis, error) {
+	expr, err := parser.ParseExpr(query)
+	if err != nil {
+		return nonShardableQuery(), err
+	}
+
+	isShardable := true
+	var analysis QueryAnalysis
+	parser.Inspect(expr, func(node parser.Node, nodes []parser.Node) error {
+		switch n := node.(type) {
+		case *parser.SubqueryExpr:
+			isShardable = false
+			return fmt.Errorf("expressions with subqueries are not shardable")
+		case *parser.Call:
+			if n.Func != nil && contains(n.Func.Name, nonShardableFuncs) {
+				isShardable = false
+				return fmt.Errorf("expressions with %s are not shardable", n.Func.Name)
+			}
+		case *parser.BinaryExpr:
+			analysis = analysis.scopeToLabels(n.VectorMatching.MatchingLabels, n.VectorMatching.On)
+		case *parser.AggregateExpr:
+			labels := make([]string, 0)
+			if len(n.Grouping) > 0 {
+				labels = n.Grouping
+			}
```

Comment:
I think we can simplified line 56 ~ 59 to just `labels := n.Grouping`, or simply pass `n.Grouping` to scopeToLabels